### ISSUE TITLE
Help with migration

### ISSUE THREAD

**Issue by revolter** [fabric/fabric#1857]
Could someone please help me migrate [this fabfile](https://github.com/revolter/DexRoBot/blob/master/fabfile.py) from version 1 to 2?

I tried to follow the migration guide but its incomplete, and it's all very very overwhelming. I already have 20 opened tabs and I'm unable to figure out how and what's the best way to do things. There are 3 different projects, with their own docs, configurations can be done from either of the projects, and in multiple different ways (config file, env vars, etc) and StackOverflow is full of old answers for the Fabric 1.

I currently have this:

```python
import configparser
import sys

from fabric import task
from invoke import env

try:
    config = configparser.ConfigParser()

    config.read('fabfile.cfg')

    env.hosts = [config.get('Fabric', 'Host')]
    env.user = config.get('Fabric', 'User')
    env.key_filename = config.get('Fabric', 'KeyFilename')

    env.project_path = config.get('Fabric', 'ProjectPath')
except configparser.Error as error:
    print('Config error: {}'.format(error))

    sys.exit(1)

env.project_name = 'DexRoBot'
env.source_filenames = [
    'main.py',
    'database.py',
    'utils.py',
    'analytics.py',
    'constants.py',

    'config.cfg',

    'requirements.txt'
]


@task(hosts=env.hosts)
def setup(context):
    context.run('mkdir -p {0.project_path}/{0.project_name}'.format(env))
    context.run('ln -s {0.project_path}/{0.project_name} ~/{0.project_name}'.format(env))

    context.run('cd ~/{.project_name}'.format(env))
    context.run('virtualenv -p python3 env')


@task
def cleanup(context):
    context.run('rm -r ~/{.project_name}'.format(env))
    context.run('rm -r {0.project_path}/{0.project_name}'.format(env))


@task(default=True, help={'filename': 'An optional filename to deploy to the server'})
def deploy(context, filename=None):
    context.run('cd ~/{.project_name}'.format(env))

    if not filename:
        for source_filename in env.source_filenames:
            context.put('src/{}'.format(source_filename), '~/{.project_name}/'.format(env))

        context.run('source env/bin/activate; pip install -r requirements.txt')
    else:
        context.put('src/{}'.format(filename), '~/{.project_name}/'.format(env))


@task
def execute(context, command=None):
    if not command:
        return

    context.run(command)
```

which doesn't work.

I don't know and can't find how to:
1. Specify the host for all the tasks without adding it to all the `@task`s
2. Specify the user
3. Specify the key_filename

And if is the `env` from `invoke` used anymore, and if not, why wasn't it removed.

**Comment by revolter** [comment]
I managed to get it working by adding:

```python
@task
def config(context):
    context.user = env.user
    context.connect_kwargs = {'key_filename': os.path.expanduser(env.key_filename)}


@task(pre=[config], hosts=env.hosts)
def setup(context):
    ...
```

But:
1. The `env.key_filename` contains a path that begins with a tilde, and it was working in Fabric 1. Was automatic path expansion removed in Fabric 2?
2. Isn't there a simpler way to do this? As I'd have to add the `config` task as a pre-task to all the rest of the tasks.

**Comment by revolter** [comment]
The entire script is functional now:

```python
import configparser
import os
import sys

from fabric import task
from invoke import env

try:
    config = configparser.ConfigParser()

    config.read('fabfile.cfg')

    env.hosts = [config.get('Fabric', 'Host')]
    env.user = config.get('Fabric', 'User')
    env.key_filename = config.get('Fabric', 'KeyFilename')

    env.project_path = config.get('Fabric', 'ProjectPath')
except configparser.Error as error:
    print('Config error: {}'.format(error))

    sys.exit(1)

env.project_name = 'DexRoBot'
env.source_filenames = [
    'main.py',
    'database.py',
    'utils.py',
    'analytics.py',
    'constants.py',

    'config.cfg'
]
env.meta_filenames = [
    'Pipfile',
    'Pipfile.lock'
]


@task
def config(context):
    context.user = env.user
    context.connect_kwargs = {'key_filename': os.path.expanduser(env.key_filename)}


@task(pre=[config], hosts=env.hosts, help={'command': 'The shell command to execute on the server'})
def execute(context, command=None):
    if not command:
        return

    context.run(command)


@task(pre=[config], hosts=env.hosts)
def cleanup(context):
    context.run('rm -rf {.project_name}'.format(env))
    context.run('rm -rf {0.project_path}/{0.project_name}'.format(env))


@task(pre=[config, cleanup], hosts=env.hosts)
def setup(context):
    context.run('mkdir -p {0.project_path}/{0.project_name}'.format(env))
    context.run('ln -s {0.project_path}/{0.project_name} {0.project_name}'.format(env))

    context.run('python -m pip install --user pipenv')


@task(default=True, pre=[config], hosts=env.hosts, help={'filename': 'An optional filename to deploy to the server'})
def deploy(context, filename=None):
    if not filename:
        for source_filename in env.source_filenames:
            context.put('src/{}'.format(source_filename), '{.project_name}/'.format(env))

        for meta_filename in env.meta_filenames:
            context.put(meta_filename, '{.project_name}/'.format(env))

        with context.cd(env.project_name):
            context.run('python -m pipenv install --three')
    else:
        context.put('src/{}'.format(filename), '{.project_name}/'.format(env))
```

But:
1. I still didn't find any information regarding the `env` from `invoke`, if I should stop using it or not. And if not, if I'm using it correctly, as for now, I'm just using it to pass data and to easily format strings.
2. I don't know how to avoid specifying the `hosts` multiple times.
3. I don't know how to get rid of the `config` task by:
    - Specifying the `user` only once and for it to be loaded automatically by all the tasks.
    - Replacing `context.connect_kwargs = {'key_filename': os.path.expanduser(env.key_filename)}` with a specific API for the SSH key.
4. I don't know how to call a task (`execute`) from another task.
5. I don't know if logging is available so you know the state and progress of a running task.

**Comment by bitprophet** [comment]
No action to take here (tho a lot of these are covered in the upgrade doc or other tickets - fwiw!) so closing. Not locking discussion though. (Sadly I don't have time to go in depth here but somebody else may.)

**Comment by vinodc** [comment]
>  I already have 20 opened tabs and I'm unable to figure out how and what's the best way to do things. There are 3 different projects, with their own docs, configurations can be done from either of the projects, and in multiple different ways (config file, env vars, etc) and StackOverflow is full of old answers for the Fabric 1.

This has been my experience as well, unfortunately. I'm also having trouble porting a v1 fabfile.
 
> I don't know how to call a task (execute) from another task.

For any one else looking into this, you might be able to try this patch: https://github.com/pyinvoke/invoke/pull/613/files
